tripy/tests/spec_verification/test_dtype_constraints.py

```diff
@@ -17,118 +17,188 @@
+                for name_not_equal in negative_test_dtypes.keys():
+                    if name_temp != name_not_equal:
+                        temp_dict[name_not_equal] = all_dtypes
                 dtype_lists_list.append(temp_dict)
             case_name = "invalid: "
         for dtype_lists in dtype_lists_list:
             for combination in itertools.product(*(dtype_lists.values())):
-                # Create a tuple with keys and corresponding elements
+                # Create a tuple with keys and corresponding elements.
                 namespace = dict(zip(dtype_lists.keys(), map(lambda val: getattr(tp, val), combination)))
                 ids = [f"{dtype_name}={dtype}" for dtype_name, dtype in namespace.items()]
-                func_list.append((func_obj, inputs, returns, namespace, positive_case, case_name + ", ".join(ids)))
+                func_list.append(
+                    (
+                        func_name,
+                        func_obj,
+                        inputs,
+                        return_dtype,
+                        namespace,
+                        positive_case,
+                        func_name + "_" + case_name + ", ".join(ids),
+                    )
+                )
 
 
-@pytest.mark.parametrize("test_data", func_list, ids=lambda val: val[5])
-def test_dtype_constraints(test_data):
-    func_obj, inputs, returns, namespace, positive_case, _ = test_data
-    # create all inputs for positive test case
+def _run_dtype_constraints_subtest(test_data):
+    func_name, func_obj, inputs, _, namespace, _, _ = test_data
     kwargs = {}
+    # Create all input objects using object_builders.create_obj.
     for param_name, input_desc in inputs.items():
-        kwargs[param_name] = create_obj(func_obj, param_name, input_desc, namespace)
-    # run api call
+        kwargs[param_name] = create_obj(param_name, input_desc, namespace)
+    # Run api call through _method_handler and setup namespace (api_call_locals).
     api_call_locals = {"kwargs": kwargs}
-    exec(method_handler(kwargs, func_obj), globals(), api_call_locals)
-    # assert with output
+    _method_handler(func_name, kwargs, func_obj, api_call_locals)
+    # If output does not have dtype skip .eval().
+    if isinstance(api_call_locals[RETURN_VALUE], int): 
+        return api_call_locals, namespace
+    # If output is a list then checking the return the first element in the list. (Assumes list of Tensors)
+    if isinstance(api_call_locals[RETURN_VALUE], List): 
+        api_call_locals[RETURN_VALUE] = api_call_locals[RETURN_VALUE][0]
+    # Run eval to check for any backend errors.
+    api_call_locals[RETURN_VALUE].eval()
+    return api_call_locals, namespace
+
+
+@pytest.mark.parametrize("test_data", func_list, ids=lambda val: val[6])
+def test_dtype_constraints(test_data):
+    _, _, _, return_dtype, _, positive_case, _ = test_data
     if positive_case:
```

Comment:
You can use `contextlib.ExitStack` to simplify this logic. You can merge the branches and then do:
```
if not positive_case:
    stack.enter_context(helper.raises(TripyException))
```
at the top.